plugins/org.eclipse.fordiac.ide.typemanagement/src/org/eclipse/fordiac/ide/typemanagement/wizards/RestrictedNewTypeWizard.java

```diff
@@ -0,0 +1,104 @@
+import org.eclipse.jface.viewers.TableViewer;
+import org.eclipse.jface.window.Window;
+import org.eclipse.jface.wizard.WizardDialog;
+import org.eclipse.swt.widgets.Composite;
+import org.eclipse.swt.widgets.Control;
+import org.eclipse.swt.widgets.Tree;
+import org.eclipse.swt.widgets.TreeItem;
+import org.eclipse.ui.PlatformUI;
+
+public class RestrictedNewTypeWizard extends NewTypeWizard {
+
+	private RestrictedNewTypeWizard() {
+	}
+
+	@Override
+	protected NewFBTypeWizardPage createNewFBTypeWizardPage() {
+		return new NewFBTypeWizardPage(new StructuredSelection());
+	}
+
+	public static TypeEntry showRestrictedNewTypeWizard(final String name, final String fileEnding,
+			final String projectName) {
+		final RestrictedNewTypeWizard wizard = new RestrictedNewTypeWizard();
+		final WizardDialog dialog = new WizardDialog(PlatformUI.getWorkbench().getActiveWorkbenchWindow()
+				.getActivePage().getActivePart().getSite().getShell(), wizard);
+		dialog.create();
+
+		final NewFBTypeWizardPage page = (NewFBTypeWizardPage) dialog.getCurrentPage();
+		page.setTemplateFileFilter(pathname -> pathname.getName().toUpperCase().endsWith(fileEnding.toUpperCase()));
+		page.setFileName(name);
+
+		final TableViewer tv = page.getTemplateViewer();
+		final Object selection = tv.getElementAt(0);
+
+		// have to check if any template at all could be found
+		if (selection != null) {
+			tv.setSelection(new StructuredSelection(selection), true);
+			// unlock table if there are possible selections
+			if (tv.getElementAt(1) == null) {
+				tv.getTable().setEnabled(false);
+			}
+		}
+		if (projectName != null) {
+			restrictProjectSelection(page.getControl(), projectName);
+		}
+		if (Window.OK == dialog.open()) {
+			// the type could be created new execute the command
+			final TypeEntry newEntry = wizard.getTypeEntry();
+			newEntry.refresh();
+			return newEntry;
+		}
+		return null;
+	}
+
+	private static void restrictProjectSelection(final Control control, final String projectName) {
+		final Composite root = (Composite) control;
+		final List<Control> elements = getFlattenedList(root);
+		//@formatter:off
+		elements.stream()
+			.filter(Tree.class::isInstance)
+			.map(Tree.class::cast) //get the tree
+			.flatMap(a-> Arrays.stream(a.getItems())) //get all root elements
+			.filter(i -> !i.getText().equals(projectName)) //find all elements which are not the current project
+			.forEach(TreeItem::dispose); //dispose them, so only the current project is left
+		//@formatter:on
+		control.redraw();
+	}
+
+	private static List<Control> getFlattenedList(final Composite root) {
+		final List<Control> flattenedSet = new ArrayList<>();
+		if (root.getChildren().length == 0) {
+			flattenedSet.add(root);
+			return flattenedSet;
+		}
+		for (final Control c : root.getChildren()) {
+			if (c instanceof final Composite com) {
+				flattenedSet.addAll(getFlattenedList(com));
+			} else {
+				flattenedSet.add(c);
+			}
+		}
+		return flattenedSet;
+	}
```

Comment:
This reaches *very* far into the internals of the wizard page up into the `TreeViewer` itself and thus seems quite messy and fragile. It looks like there is no straignforward way to filter containers in the existing wizard page base class `WizardNewFileCreationPage`.

I believe we may either have to rely on validation here or write an entirely custom wizard page, similar to `org.eclipse.jdt.ui.wizards.NewClassWizardPage`. What do you think @oberlehner @azoitl?